### PR TITLE
fix(XHRImpl): file:/// and IE9 bugs

### LINKED ISSUE #2219: templateUrl usage with local files (file:///)  not working because of (XMLHttpRequest)  xhr.status checks
- Overview of the Issue    
  templateUrl is not able to load local files (file:///) do to xhr.status (XMLHttpRequest) checks. This prevents the usage of `templateUrl` within `@Views`   
 **stack trace:**
```
Failed to load the template "MyAppComponent"

Error
    at new BaseException (https://code.angularjs.org/2.0.0-alpha.25/angular2.dev.js:7159:25)
    at https://code.angularjs.org/2.0.0-alpha.25/angular2.dev.js:22973:21
    at Zone.run (https://code.angularjs.org/2.0.0-alpha.25/angular2.dev.js:115:17)
    at Zone.execute.NgZone._createInnerZone.zone.fork.fork.$run [as run] (https://code.angularjs.org/2.0.0-alpha.25/angular2.dev.js:12418:38)
    at zoneBoundFn (https://code.angularjs.org/2.0.0-alpha.25/angular2.dev.js:88:19)
    at lib$es6$promise$$internal$$tryCatch (https://code.angularjs.org/2.0.0-alpha.25/angular2.dev.js:479:16)
    at lib$es6$promise$$internal$$invokeCallback (https://code.angularjs.org/2.0.0-alpha.25/angular2.dev.js:491:17)
    at lib$es6$promise$$internal$$publish (https://code.angularjs.org/2.0.0-alpha.25/angular2.dev.js:462:11)
    at lib$es6$promise$$internal$$publishRejection (https://code.angularjs.org/2.0.0-alpha.25/angular2.dev.js:412:7)
    at https://code.angularjs.org/2.0.0-alpha.25/angular2.dev.js:195:9

```    

- Motivation for or Use Case      
Angular in combination with Electron must be able to load local files (file:///) like `"file:///path/to/myAppComponent.html"`. The issue appears when using `templateUrl` instead of `template` within a view. The xhr implementation expects a status response `(200 <= status && status <= 300)`, but for local files, browsers do not respond with a status code (0). Have a look at [xhr_impl.ts L15](https://github.com/angular/angular/blob/91ccc9af9846b2ce54abe2ca8fff149438c57798/modules/angular2/src/services/xhr_impl.ts#L15). This issue makes the usage of template files impossible when developing inside electron applications.    
- Angular Version(s)    
 Tested with 2.0.0-alpha.25
- Browsers and Operating System     
 all
- Reproduce the Error     
  - http://plnkr.co/edit/1P2LpY?p=info
  - Fork the repository
  - Run electron `./electron bug/`

### PR DESCRIPTION
**Pull request by PatrickJS**
closes #2219

### CODE REVIEW

File: modules/angular2/src/services/xhr_impl.ts [hunk 1]
```diff
@@ -1,19 +1,33 @@
...
 import {Promise, PromiseWrapper} from 'angular2/src/facade/async';
+import {global} from 'angular2/src/facade/lang';
 import {XHR} from './xhr';
 
 @Injectable()
 export class XHRImpl extends XHR {
   get(url: string): Promise<string> {
     var completer = PromiseWrapper.completer();
-    var xhr = new XMLHttpRequest();
+    var xhr = new global.XMLHttpRequest();
```
Comment (vicb) [inline modules/angular2/src/services/xhr_impl.ts]: Why ?

**Comment by vicb** [comment]
Needs a test, dors thé Dart implémentation works with file ?

**Comment by vicb** [comment]
Any update on the tests ?

**Comment by vicb** [comment]
@gdi2290 any updates on the tests ?

**Comment by gdi2290** [comment]
yeah, let me rebase and add tests

**Comment by vicb** [comment]
Thanks for adding tests

**Comment by gdi2290** [comment]
@vicb actually I'm not sure how to tests for these new cases. It's not like I could mock out the backend to return status `0`, status `1223`, or respond in `responseText`. And everywhere I look it seems like no one (including the angular.js repo) is testing for file://

**Comment by vicb** [comment]
Could you check angular 1, see if there are tests. Thanks

**Comment by gdi2290** [comment]
@vicb I found tests for [httpBackendSpec.js#L360](https://github.com/angular/angular.js/blob/master/test/ng/httpBackendSpec.js#L360). For Angular1 there is a way to mutate or change the XMLHttpRequest instance during testing. The current implementation for Angular2 was meant to be extremely minimal which is why it only does a `get` request. For example, there is no way for me to test [1223 status code](https://github.com/angular/angular.js/blob/master/test/ng/httpBackendSpec.js#L94) since I'm unable to manually mutate the status to `1223` and test the `onload` function. Without a refactor to either allow me to intercept the XMLHttpRequest instance or provide one as an option as a constructor argument I'm still having a hard time testing this. Searching online for the error currently in travis I found that many people suggest turning on flags such as `--allow-file-access-from-files` or `--web-security` in order for file:// to be used with xhr which doesn't seem like a great option. In Angular1, this was largely avoided since they're able to grab the xhr instance to mutate any values needed and manually run `onload` in order to determine the correct results. Then there's Dart which I'm not an expert on to know what applies in js vs dart's XMLHttpRequest. I also still have a gulp build problem with broccoli that other people still have so I'm unable to test anything locally :/ So I need some direction again unless I overlooked something.


Currently there are 3 things that should be tested since they were added in this PR

* should work with either `xhr.response` or `xhr.responseText`
* should normalize status code for IE9 1223 bug
* should fix status code when it is 0 for files

**Comment by vicb** [comment]
I see little value in mocking the `XmlHttpRequest`, we want to test with a real one (IMO if we write the test and tested object then it is no more a test).

Dart has a different implementation.

Could you remove the failing and I'll merge the PR as is.

**Comment by gdi2290** [comment]
@vicb it should be good to merge now

**Comment by vicb** [comment]
@gdi2290 thanks and btw what is your problem to test locally ? we should try to fix that

**Comment by gdi2290** [comment]
@vicb here's an image
screen shot 2015-06-16 at 11 32 52 pm
@PascalPrecht also ran into the problem which is probably why zippy wasn't refactored https://github.com/angular/angular/pull/729#issuecomment-107903875
There was another issue somewhere where it was reported but I can't remember where it was

**Comment by vicb** [comment]
That should be an easy one, rm the `typings` fodlers in `module/angular2` and `tools`, run `npm i` and you should be good (do not hesitate to create GH issues for build pbs).

**Comment by gdi2290** [comment]
awesome that worked! I wish the error message was more clear on what's going on but at least I can build now. thanks

**Comment by vicb** [comment]
landed as cd735c483747c66d378ae34f4755178dd0a54d70, thanks.

**Pull request closed without merging**